processes/ingest/clacso.py

```diff
@@ -1,45 +1,106 @@
             elif self.process == 'complete':
-                records = self.clacso_service.get_records(full_import=True, offset=self.offset, limit=self.limit)
+                records = self.dspace_service.get_records(full_import=True, offset=self.offset, limit=self.limit)
             elif self.process == 'custom':
-                records = self.clacso_service.get_records(start_timestamp=self.ingestPeriod, offset=self.offset, limit=self.limit)
+                records = self.dspace_service.get_records(start_timestamp=self.ingestPeriod, offset=self.offset, limit=self.limit)
             else: 
                 logger.warning(f'Unknown CLACSO ingestion process type {self.process}')
                 return
-            for record in records:
-                self.addDCDWToUpdateList(record)
             
+            if records:
+                for record in records:
+                    self.store_pdf_manifest(record)
+                    self.addDCDWToUpdateList(record)
+
             self.saveRecords()
             self.commitChanges()
 
-            logger.info(f'Ingested {len(records)} CLACSO records')
+            logger.info(f'Ingested {len(self.records) if records else 0} CLACSO records')
 
         except Exception as e:
             logger.exception('Failed to run CLACSO process')
             raise e   
         finally:
             self.close_connection()
+
+    def store_pdf_manifest(self, record):
+        for link in record.record.has_part:
+            item_no, uri, source, media_type, flags = link.split('|')
+
+            if media_type == 'application/pdf':
+                record_id = record.record.identifiers[0].split('|')[0]
+
+                manifest_path = 'manifests/{}/{}.json'.format(source, record_id)
+                manifest_uri = 'https://{}.s3.amazonaws.com/{}'.format(
+                    self.s3_bucket, manifest_path
+                )
+
+                manifest_json = self.generate_manifest(record.record, uri, manifest_uri)
+
+                self.s3_manager.createManifestInS3(manifest_path, manifest_json, self.s3_bucket)
+
+                link_string = Part(
+                    index=item_no, 
+                    url=manifest_uri,
+                    source=source,
+                    file_type='application/pdf',
+                    flags=json.dumps(dataclasses.asdict(FileFlags()))
+                ).to_string()
+                record.record.has_part.insert(0, link_string)
+                break
+
+    @staticmethod
+    def generate_manifest(record, sourceURI, manifest_urI):
+        manifest = WebpubManifest(sourceURI, 'application/pdf')
+
+        manifest.addMetadata(record)
+        
+        manifest.addChapter(sourceURI, record.title)
+
+        manifest.links.append({
+            'rel': 'self',
+            'href': manifest_urI,
+            'type': 'application/webpub+json'
+        })
+
+        return manifest.toJson()
```

Comment:
what are your thoughts on pulling these functions out? I think we have very similar if not the same code in several ingest processes?

Reply:
I agree with pulling these methods out into the S3Manager like the `createManifestInS3` method is already. I'll make another PR to pull these methods into the S3Manager and change all the ingest processes to reflect the change.